### DISCUSSION TITLE
Is there a way to center the models that works generally?

### DISCUSSION THREAD

**Discussion by jp06** [flowtsohg/mdx-m3-viewer#61] [Q&A]
Gonna post this here so that I don't muddy the Issues section.

So I wanted to make the model zoomed enough but without overlapping the canvas as much as possible. So far, my go to was to `instance.setLocation([0, 0, -75])` (in my case) and it looks centered for the most part but there are some models that are either a bit too low/high or too tall, which makes me set the camera `distance` to `400` to be able to fully see those also (else I'd go like `300` or `325`). 

Seeing that the initial camera location seems to be centered at the bottom of the model, I've thought about setting the camera z coordinate to something like `-(model_height / 2)` instead of a fixed one. Except, I'm helpless and can't figure out how to get that height. Need help regarding getting the model height or any alternate solution that can consistently center the model in the viewer. Thank you.

**Comment by flowtsohg** [comment]
Every model has a `bounds` member, which defines both a bounding sphere and AABB.
Practically speaking right now for MDX models, the bounds is initialized with the MDX's own bounds.

For many units this is not a tight fit, because for example it might be a lot bigger than the model if its animations are much bigger than it.
E.g. think of archer units - they typically have arrows shooting out in their attack animations up to some distance (not the in-game projectile, but a fake one in the animation itself, before the game spawns the projectile), and their bounds include the arrow mesh when its the furthest.
This is because these bounds are used for culling when rendering, not for game logic like collisions and projectiles and whatnot.

Like everything WC3 related, it's a bit hard to actually define what a tight fit would be, because it's not obvious what mesh is even relevant.
In the archer example, should the arrow mesh really affect the size of the archer? probably not, but how would the code know that.
Should team glows affect it? probably not, but how would the code know that (guess based on the texture and number of vertices in the mesh? I've done that in the past for color-picking and the results were generally good)
Should the code calculate the per-geoset bounding shapes for every model now which is kind of expensive at load-time, given that many models don't have them and they seem completely optional (or are not even used by the game in the first place?)

You could get the per-animation bounds and they might be less big, e.g. use a "Stand" animation which will avoid the issues of archers and arrows (or whatever other similar model configurations), but many models don't have them, or they are not actually the right size, and we get back to all these questions and inconsistencies.

There are also models with no meshes at all, e.g. pure emitters, but we can call those special cases and ignore them for now.

So...it should be a really simple question to answer, but I am not actually sure what's the best way to get a model's height, or its size in general.

**Comment by flowtsohg** [comment]
I finally did the experiment of using the per-animation extents, which I thought of doing years ago.
Now I feel silly for not doing it back then.

For most models these actually seem like a very good estimation of both size and location (e.g. in the case of portraits that are not centered).
For the game models I checked, it seems to be more or less perfect.
Get the animation that the model is most likely to be in its most basic rest pose, probably always the first Stand animation(?), and use its extent as a bounding box/sphere, whose size and location are the estimated size and location of the model.

This is still Warcraft 3, and if I learned one thing in all these years, it's that there is not one consistent thing in Warcraft 3.
There are models without/with wrong main extents.
There are models with sequences without/with wrong extents.
There are models with no sequences at all.

With that said, I can't think of any other option, and honestly the results were so good for most models I tested, custom ones included, that I am thinking about starting to use this like I planned to all that time ago - culling instances based on the running animation. Should remove many false-positives when checking for visibility, mostly for the map viewer.

I secretly added code to the sanity test client on the hive to show the extents visually. I'll probably commit it later.
If you want to check how fitting it is for models relevant to you, [it's here](https://viewer.hiveworkshop.com/check/) (drag & drop your models / custom textures anywhere on the page).
If running no animation, or the current animation has no extent, the model extent will be shown.

**Reply by jp06** [reply]
Weird, I thought I replied to this thread days ago, I think it didn't actually send through. I've been playing with that value when you suggested it previously. Back then, I just tried getting a stand sequence and then setting the location z to `(extent.min[2] - extent.max[2]) / 2` with the stand sequence's extent. It ended up more inconsistent because I can't figure out the math so I dropped the idea and gave up on that. Now thanks to those box thingies in the viewer you linked, I realized that some stand models actually have extents that are too big and might not necessarily be the smallest ones. I'll come back to it again and this time I'll go through all the sequences (I don't know why I didn't do that in the first place 🤦‍♂️).

**Reply by jp06** [reply]
I still wasn't able to get it right, it seems that it's not that simple 😅. 

It doesn't work on those that have big bounds around the model (when I check on the viewer you linked on those not centered). I also ended up excluding the decay sequences because they have much lower height than the stand sequences. I noticed along the way that there are models that have some very large extents (goes into scientific notation up to e38). It ends up getting those so I had to exclude when it gets to a big enough number (I checked for 1e4 and above but there must be a better way to check). 

I forgot to say that all the models I opened in the viewer are centered well within the bounds (if that's what you're asking).

**Comment by flowtsohg** [comment]
Initially I thought it could be any animation, and to simply get the smallest extent among them, but then I realized that's not really logically correct.
Really what we want is the base pose of the model, which I think will always be the stand animation with smallest extent?

Some models definitely will have huge extents - modders use that to ensure their models are not culled, especially for models that are planned to be gigantic in the game (like World Tree gigantic) and you want them to render for sure and not disappear when the camera is moved a bit.

Here's the code in the sanity test client for reference:
```ts
let extent;

if (index === -1) {
  extent = this.visibleTest.parser.extent;
} else {
  extent = this.visibleTest.parser.sequences[index].extent;
}

let { max, min } = extent;
let x = (max[0] + min[0]) / 2;
let y = (max[1] + min[1]) / 2;
let z = (max[2] + min[2]) / 2;
let w = (max[0] - min[0]) / 2;
let d = (max[1] - min[1]) / 2;
let h = (max[2] - min[2]) / 2;

this.visibleTest.boundingBox.setLocation([x, y, z]);
this.visibleTest.boundingBox.setScale([w, d, h]);

this.visibleTest.boundingSphere.setLocation([x, y, z]);
this.visibleTest.boundingSphere.setUniformScale(extent.boundsRadius);
```
Note that the sizes are divided by 2 because the unit sized primitives are size 2 (-1 to 1). Maybe I should finally make them legit size 1 (-0.5 to 0.5) and all scaling will be more intuitive...
Either way, if you want the actual size of the extent, don't divide by 2.

**Reply by jp06** [reply]
I actually was dividing the difference of the two extents by 2 only for using it as an offset. I had the assumption that the center is base of the models, but I noticed that those flying units are way below the center for some reason. I thought setting the location to that would center it, but there are some models like Beastmaster that has a bigger stand animation bound for its size but is also centered on around its feet on it's initial location in the viewer. (I think) I solved the flying units problem which I'll post as an answer below because it has solved my problem, but I'm still using a static offset for the location in the z-axis.

**Comment by jp06** [comment]
By the way, I just only noticed the `unit.ts` on your w3x parser a while ago. It would have saved a few headaches from trying to figure out the correct scaling for the heroes had I discovered it earlier 🤦‍♂️. When I saw you using that `moveHeight` property of the unit, it hit me that it might be the missing piece to "normalizing" the base location of the flying units. Thank you for that!

Now with `instance.setLocation([0, 0, -100 + (moveHeight ? moveHeight / 2 : 0)])`, I finally was able to correct the base of the flying units and zoom in the `distance` of `325` with the biggest model I think I have (Centaur Khan) occupying the canvas while having all other models not getting culled. `moveHeight / 2` will only be added to those that have the property which are mostly flying units. This might not reliably work for everyone but hopefully for those having a problem with the flying units being too low, this might give you an idea how to solve it.

![finally](https://user-images.githubusercontent.com/30029476/107083423-91df9500-6830-11eb-93a4-d1c722e2331b.png)

**Comment by flowtsohg** [comment]
If you are working in the context of a map, yeah the map data might help.

I tested more models.
Every game model I tried would work more or less perfectly if the extents location is used to center the model.
When it comes to custom models, it's more questionable and less useful. Some have completely wrong extents, but even for those that probably used Magos or some tool to calculate them, they are all the same as the model extents and nothing is gained by using the per-animation ones.

Back to not knowing if there's a proper way to get the size of a model :/
Obviously these extents can be generated automatically and always fit, but that is a lot of code and a lot of processing.

I can still probably use them for the viewer, because if they are buggy, they would be buggy in the game as well presumably, so it's...intended buggy behavior.

**Reply by jp06** [reply]
I actually only used the map parser to get the name of the map and load certain files to get some data and output it in a JSON file which contains the units' model, art and other properties, so it never occurred to me to check the stuff underneath too much 😅. I extracted the models and textures of the map manually to the folder with it's version as the name. And then I load the models and textures from the server, resolving from DotA map resources, patch resources, TFT resources, and finally the ROC resources before returning a 404 if resource is not found in all them. Although there are some hidden files like `Objects\Spawnmodels\Human\HumanBlood\HumanBloodSmall0.mdx` that I don't know how to extract which is the only error I get in the console now (but I'm not sure how I managed to extract `ReplaceableTextures\CommandButtons\BTNElunesBlessing.blp` back then).

**Comment by flowtsohg** [comment]
Do you mean the above file fails to decode? Not sure if you are talking about my code or something else. If my code, what error do you mean?

**Reply by jp06** [reply]
Ah no, I mean it's just that I can't see it with Ladik's MPQ editor in all the MPQs I checked.  It's something on my side only because I can't find it.

**Comment by flowtsohg** [comment]
BTNElunesBlessing.blp is in War3x.mpq, and HumanBloodSmall0.mdx doesn't exist as far as I know 🤔

**Reply by jp06** [reply]
These spawn models are the ones I can't retrieve:
* `Objects\Spawnmodels\NightElf\NightElfBlood\NightElfBloodHeroMoonPriestess.mdx`
* `Objects\Spawnmodels\Human\HumanBlood\HumanBloodSmall0.mdx`
* `Objects\Spawnmodels\Orc\Orcblood\OrcBloodRiderlessWyvernRider.mdx`

From these models respectively:
* `units\nightelf\HeroMoonPriestess\HeroMoonPriestess.mdx`
* `units\human\Sorceress\Sorceress.mdx`
* `units\orc\SpiritWyvern\SpiritWyvern.mdx`

Although, it seems like on the Riderless Wyvern's case, it's that the actual spawn model file has a typo (Or**d**Blood instead of Or**c**Blood). I actually only noticed by putting it on a string difference checker, it's about to drive me crazy wondering why it always return a 404. It doesn't really cause me any problems so I just ignore those missing ones.

How did you extract/view `BTNElunesBlessing.blp` though? I can't find it in the MPQ editor.

**Reply by jp06** [reply]
I just checked in the online viewer and it's the same.

On a bit of unrelated note, I just noticed that Hive has an API for the TFT models. Is it free for use for the public? Like if you are to use it as the server for your models and textures?

**Comment by flowtsohg** [comment]
You are not the only one [annoyed by Blizzard patches](https://github.com/flowtsohg/mdx-m3-viewer/blob/master/clients/tests/index.js#L9), my friend :)
The path used to be fine until they started pushing patches before Reforged, I don't remember when exactly they broke this specific path.
I just noticed it because it happened to be used by one of the unit tests.

`BTNElunesBlessing.blp` is in `War3x.mpq`, at least it is in `1.29.2`.

`HumanBloodSmall0.mdx` I couldn't find anywhere as well. It might be a relic of the past, there are many leftovers in WC3 data all over the place (or even in the nearer past by breaking it in some patch).

`NightElfBloodHeroMoonPriestess.mdx` I can't find either, but I did notice `NightElfBloodMoonPriestess.mdx` (without the `Hero`) in `War3.mpq`.
A bad un-updated reference in the model? There are many of those as well in WC3 data.
Would be funny to go in-game and kill a POTM, and realize no blood comes out of her when she dies because of this, but it's too much effort for me to bother checking.

Hive has now APIs for both TFT and Reforged. They are technically(?) not private, as they are [right here](https://github.com/flowtsohg/mdx-m3-viewer/blob/master/clients/shared/localorhive.js). If you don't abuse them too much it will probably be ok :)
There are still some CORS issues sometimes, but generally they work.
Do note that the 404 page also results in a CORS error, which many times can be confusing because you'd expect a 404 error if a file can't be found. It is what it is 🤷
